types/rlp_unmarshal.go

```diff
@@ -301,86 +301,131 @@ func (t *Transaction) UnmarshalRLP(input []byte) error {
-	}
+	var err error
 
 	p.Hash(t.Hash[:0], v)
 
+	// Setup defaults
+	t.GasPrice = new(big.Int)
+	t.GasFeeCap = new(big.Int)
+	t.GasTipCap = new(big.Int)
+	t.Value = new(big.Int)
+	t.V = new(big.Int)
+	t.R = new(big.Int)
+	t.S = new(big.Int)
+	t.From = ZeroAddress
+
+	// Type
+	i := 0
+	if t.Type, err = ReadRlpTxType(v.Get(i)); err != nil {
+		return err
+	}
+
 	// nonce
-	if t.Nonce, err = elems[0].GetUint64(); err != nil {
+	i++
+	if t.Nonce, err = v.Get(i).GetUint64(); err != nil {
 		return err
 	}
+
 	// gasPrice
-	t.GasPrice = new(big.Int)
-	if err := elems[1].GetBigInt(t.GasPrice); err != nil {
+	i++
+	if err = v.Get(i).GetBigInt(t.GasPrice); err != nil {
 		return err
 	}
+
+	if t.Type == DynamicGeeTx {
+		// gasFeeCap
+		i++
+		if err = v.Get(i).GetBigInt(t.GasFeeCap); err != nil {
+			return err
+		}
+
+		// gasTipCap
+		i++
+		if err = v.Get(i).GetBigInt(t.GasTipCap); err != nil {
+			return err
+		}
+	}
+
 	// gas
-	if t.Gas, err = elems[2].GetUint64(); err != nil {
+	i++
+	if t.Gas, err = v.Get(i).GetUint64(); err != nil {
 		return err
 	}
+
 	// to
-	if vv, _ := v.Get(3).Bytes(); len(vv) == 20 {
+	i++
+	if vv, _ := v.Get(i).Bytes(); len(vv) == 20 {
 		// address
 		addr := BytesToAddress(vv)
 		t.To = &addr
-	} else {
-		// reset To
-		t.To = nil
 	}
+
 	// value
-	t.Value = new(big.Int)
-	if err := elems[4].GetBigInt(t.Value); err != nil {
+	i++
+	if err = v.Get(i).GetBigInt(t.Value); err != nil {
 		return err
 	}
+
 	// input
-	if t.Input, err = elems[5].GetBytes(t.Input[:0]); err != nil {
+	i++
+	if t.Input, err = v.Get(i).GetBytes(t.Input[:0]); err != nil {
 		return err
 	}
 
 	// V
-	t.V = new(big.Int)
-	if err = elems[6].GetBigInt(t.V); err != nil {
+	i++
+	if err = v.Get(i).GetBigInt(t.V); err != nil {
 		return err
 	}
 
 	// R
-	t.R = new(big.Int)
-	if err = elems[7].GetBigInt(t.R); err != nil {
+	i++
+	if err = v.Get(i).GetBigInt(t.R); err != nil {
 		return err
 	}
+
 	// S
-	t.S = new(big.Int)
-	if err = elems[8].GetBigInt(t.S); err != nil {
+	i++
+	if err = v.Get(i).GetBigInt(t.S); err != nil {
 		return err
 	}
 
-	if len(elems) >= 10 {
-		// Type
-		if t.Type, err = ReadRlpTxType(elems[9]); err != nil {
-			return err
-		}
-
-		if t.IsStateTx() {
-			// We need to set From field for state transaction,
-			// because we are using unique, predefined address, for sending such transactions
-			// From
-			if vv, err := v.Get(10).Bytes(); err == nil && len(vv) == AddressLength {
-				// address
-				addr := BytesToAddress(vv)
-				t.From = addr
-			} else {
-				// reset From
-				t.From = ZeroAddress
-			}
+	if t.Type == StateTx {
+		// We need to set From field for state transaction,
+		// because we are using unique, predefined address, for sending such transactions
+		// From
+		i++
+		if vv, err := v.Get(i).Bytes(); err == nil && len(vv) == AddressLength {
+			// address
+			addr := BytesToAddress(vv)
+			t.From = addr
 		}
 	}
 
 	return nil
 }
+
+func ReadRlpTxType(rlpValue *fastrlp.Value) (TxType, error) {
```

Comment:
Do we need this exported?

Reply:
Done: https://github.com/0xPolygon/polygon-edge/pull/1111/commits/6d99243429dec0ab1bf3c1a3f63e4a8145941d5c